File: CRM-Server/app/schemas/operation_log.py
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List, Dict, Any
from datetime import datetime
import enum


class EventAction(str, enum.Enum):
    CREATE = "CREATE"
    UPDATE = "UPDATE"
    DELETE = "DELETE"
    SUBMIT = "SUBMIT"
    APPROVE = "APPROVE"
    REJECT = "REJECT"
    CANCEL = "CANCEL"
    CONVERT = "CONVERT"
    WIN = "WIN"
    LOSE = "LOSE"


class PrimaryResourceType(str, enum.Enum):
    LEAD = "LEAD"
    CUSTOMER = "CUSTOMER"
    OPPORTUNITY = "OPPORTUNITY"
    CONTRACT = "CONTRACT"
    INVOICE = "INVOICE"
    PAYMENT_PLAN = "PAYMENT_PLAN"
    PAYMENT_RECORD = "PAYMENT_RECORD"


class OperationLogCreate(BaseModel):
    event_type: str = Field(..., min_length=1, max_length=50, description="事件类型")
    event_action: str = Field(..., min_length=1, max_length=20, description="事件动作")
    primary_resource_type: str = Field(..., min_length=1, max_length=20, description="主资源类型")
    primary_resource_id: int = Field(..., gt=0, description="主资源ID")
    secondary_resource_type: Optional[str] = Field(None, max_length=20, description="次资源类型")
    secondary_resource_id: Optional[int] = Field(None, gt=0, description="次资源ID")
    operator_id: str = Field(..., min_length=1, max_length=100, description="操作人ID")
    operator_name: Optional[str] = Field(None, max_length=100, description="操作人姓名")
    content: Dict[str, Any] = Field(..., description="事件内容")
    remark: Optional[str] = Field(None, max_length=500, description="备注")


class OperationLogResponse(BaseModel):
    id: int = Field(..., description="记录ID")
    event_id: str = Field(..., description="事件唯一ID")
    event_type: str = Field(..., description="事件类型")
    event_action: str = Field(..., description="事件动作")
    primary_resource_type: str = Field(..., description="主资源类型")
    primary_resource_id: int = Field(..., description="主资源ID")
    secondary_resource_type: Optional[str] = Field(None, description="次资源类型")
    secondary_resource_id: Optional[int] = Field(None, description="次资源ID")
    operator_id: str = Field(..., description="操作人ID")
    operator_name: Optional[str] = Field(None, description="操作人姓名")
    operated_at: datetime = Field(..., description="操作时间")
    content: Dict[str, Any] = Field(..., description="事件内容")
    remark: Optional[str] = Field(None, description="备注")
    
    model_config = ConfigDict(from_attributes=True)


class OperationLogListResponse(BaseModel):
    list: List[OperationLogResponse] = Field(..., description="操作记录列表")
    total: int = Field(..., description="总记录数")
    page_no: int = Field(..., description="当前页码")
    page_size: int = Field(..., description="每页数量")

File: CRM-Client/src/api/operationLog.ts
import request from '@/utils/request'
import { z } from 'zod'

export type ResourceType = 
  | 'LEAD' 
  | 'CUSTOMER' 
  | 'OPPORTUNITY' 
  | 'CONTRACT' 
  | 'INVOICE' 
  | 'PAYMENT_PLAN' 
  | 'PAYMENT_RECORD'

export type EventType = 
  | 'LEAD_CREATED'
  | 'LEAD_CONVERTED'
  | 'CUSTOMER_CREATED'
  | 'MANUAL_FOLLOW_UP'
  | 'OPPORTUNITY_CREATED'
  | 'CONTRACT_CREATED'
  | 'CONTRACT_STATUS_CHANGED'
  | 'INVOICE_CREATED'
  | 'PAYMENT_RECEIVED'
  | 'SYSTEM_ALERT'

export type EventAction = 'CREATE' | 'UPDATE' | 'DELETE' | 'STATUS_CHANGE'

export type OperationLogContent = Record<string, unknown>;

export interface OperationLog {
  id: number
  event_id: string
  event_type: EventType
  event_action: EventAction
  primary_resource_type: ResourceType
  primary_resource_id: number
  secondary_resource_type: ResourceType | null
  secondary_resource_id: number | null
  operator_id: string
  operator_name: string | null
  operated_at: string
  content: OperationLogContent
  remark: string | null
}

export interface OperationLogListResponse {
  list: OperationLog[]
  total: number
  page_no: number
  page_size: number
}

export interface GetResourceLogsParams {
  primary_resource_type: ResourceType
  primary_resource_id: number
  event_types?: EventType[] | null
  page_no?: number
  page_size?: number
}

export interface GetMyLogsParams {
  page_no?: number
  page_size?: number
}

const ResourceTypeSchema = z.enum([
  'LEAD',
  'CUSTOMER',
  'OPPORTUNITY',
  'CONTRACT',
  'INVOICE',
  'PAYMENT_PLAN',
  'PAYMENT_RECORD',
])

const EventTypeSchema = z.enum([
  'LEAD_CREATED',
  'LEAD_CONVERTED',
  'CUSTOMER_CREATED',
  'MANUAL_FOLLOW_UP',
  'OPPORTUNITY_CREATED',
  'CONTRACT_CREATED',
  'CONTRACT_STATUS_CHANGED',
  'INVOICE_CREATED',
  'PAYMENT_RECEIVED',
  'SYSTEM_ALERT',
])

const EventActionSchema = z.enum(['CREATE', 'UPDATE', 'DELETE', 'STATUS_CHANGE'])

const OperationLogSchema = z.object({
  id: z.number(),
  event_id: z.string(),
  event_type: EventTypeSchema,
  event_action: EventActionSchema,
  primary_resource_type: ResourceTypeSchema,
  primary_resource_id: z.number(),
  secondary_resource_type: ResourceTypeSchema.nullable(),
  secondary_resource_id: z.number().nullable(),
  operator_id: z.string(),
  operator_name: z.string().nullable(),
  operated_at: z.string(),
  content: z.record(z.string(), z.unknown()),
  remark: z.string().nullable(),
}).passthrough()

const OperationLogListResponseSchema = z.object({
  list: z.array(OperationLogSchema),
  total: z.number(),
  page_no: z.number(),
  page_size: z.number(),
})

const operationLogApi = {
  async getResourceLogs(params: GetResourceLogsParams): Promise<OperationLogListResponse> {
    const queryParams: Record<string, unknown> = {
      primary_resource_type: params.primary_resource_type,
      primary_resource_id: params.primary_resource_id,
      page_no: params.page_no ?? 1,
      page_size: params.page_size ?? 20
    }

    if (params.event_types !== undefined && params.event_types !== null && params.event_types.length > 0) {
      queryParams['event_types'] = params.event_types.join(',')
    }

    // eslint-disable-next-line crmwolf/require-zod-schema
    const raw: unknown = await request.get('/v1/operation-logs', {
      params: queryParams
    })
    return OperationLogListResponseSchema.parse(raw)
  },

  async getMyLogs(params?: GetMyLogsParams): Promise<OperationLogListResponse> {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const raw: unknown = await request.get('/v1/operation-logs/my-logs', {
      params: params ?? {}
    })
    return OperationLogListResponseSchema.parse(raw)
  }
}

export default operationLogApi
